include/mbgl/storage/resource.hpp

```diff
@@ -20,17 +20,42 @@ class Resource {
         SpriteJSON
     };
 
-    Resource(Kind kind_, const std::string& url_)
+    struct TileData {
+        std::string urlTemplate;
+        float pixelRatio;
+        int32_t x;
+        int32_t y;
+        int8_t z;
+    };
+
+    Resource(Kind kind_, const std::string& url_, optional<TileData> tileData_ = {})
         : kind(kind_),
-          url(url_) {
+          url(url_),
+          tileData(tileData_) {
     }
 
+    static Resource style(const std::string& url);
+    static Resource source(const std::string& url);
+    static Resource tile(const std::string& urlTemplate,
+                         float pixelRatio,
+                         int32_t x,
+                         int32_t y,
+                         int8_t z);
+    static Resource glyphs(const std::string& urlTemplate,
+                           const std::string& fontStack,
+                           const std::pair<uint16_t, uint16_t>& glyphRange);
+    static Resource spriteImage(const std::string& base, float pixelRatio);
+    static Resource spriteJSON(const std::string& base, float pixelRatio);
```

Comment:
This way enforces the appropriate arguments for each kind, and is more succinct.

```
const Resource style = Resource::style(url);
```

vs

```
const Resource style(Resource::Kind::Style, url);
const Resource tile(Resource::Kind::Tile, url); // wrong arguments
```

Better would be to make `Resource` a variant type, but that's a more invasive change than I want to make right now.